### PR TITLE
Split mesonlib package to better separate os specific code

### PR DESCRIPTION
**Pull request by dcbaker**
Currently, mesonlib does some import tricks to figure out whether it
needs to use windows or posix specific functions. This is a little
hacky, but works fine. However, the way the typing stubs are implemented
for the msvcrt and fnctl modules will cause mypy to fail on the other
platform, since the functions are not implemented.

To alleviate this (and for slightly cleaner design), I've split mesonlib
into a package with three modules. A universal module contains all of
the platform agnositc code, a win32 module contains window specific
code, a posix module contains the posix specific code, and a platform
module contains no-op implementations. Then the package's __init__ file
imports all of the universal functions and all of the functions from the
appropriate platform module, or the no-op versions as fallbacks. This
makes mypy happy, and avoids `if`ing all over the code to switch between
the platform specific code.

### CODE REVIEW

File: mesonbuild/compilers/compilers.py [hunk 1]
```diff
@@ -51,7 +51,7 @@
 # This means we can't include .h headers here since they could be C, C++, ObjC, etc.
 lang_suffixes = {
     'c': ('c',),
-    'cpp': ('cpp', 'cc', 'cxx', 'c++', 'hh', 'hpp', 'ipp', 'hxx', 'ino', 'ixx'),
+    'cpp': ('cpp', 'cc', 'cxx', 'c++', 'hh', 'hpp', 'ipp', 'hxx', 'ino', 'ixx', 'C', 'H'),
```
Comment (mensinda) [inline mesonbuild/compilers/compilers.py:54]: why is this change necessary?

Comment (dcbaker) [reply mesonbuild/compilers/compilers.py:54]: Apparently I didn't git clean hard enough

**Review by mensinda — APPROVED** [review]
LGTM
